google/cloud/storage/internal/object_requests.cc

```diff
@@ -159,6 +159,49 @@ std::ostream& operator<<(std::ostream& os, UpdateObjectRequest const& r) {
   return os << "}";
 }
 
+ComposeObjectRequest::ComposeObjectRequest(std::string bucket_name,
+      std::string destination_object_name,
+      std::vector<ComposeSourceObject> source_objects,
+      ObjectMetadata destination_object_metadata)
+      : GenericObjectRequest(std::move(bucket_name),
+        std::move(destination_object_name)),
+        destination_metadata_(std::move(destination_object_metadata)),
+        source_objects_(std::move(source_objects)) {
+  using internal::nl::json;
+  json compose_object_payload_json;
+  compose_object_payload_json["kind"] = "storage#composeRequest";
+  compose_object_payload_json["destination"] = json::parse(
+    destination_metadata_.JsonPayloadForCompose());
+  json source_object_list;
+  for(auto const& source_object : source_objects_) {
+    json source_object_json;
+    source_object_json["name"] = source_object.object_name;
+    if(source_object.generation.has_value()) {
+      source_object_json["generation"] = source_object.generation.value();
+    }
+    if(source_object.if_generation_match.has_value()) {
+      source_object_json["ifGenerationMatch"] = \
+          source_object.if_generation_match.value();
+    }
+    source_object_list.push_back(std::move(source_object_json));
```

Comment:
Since you are moving, please use `source_object_list.emplace_back(std::move(source_object_json))`